lib/libxdp/libxdp.c

```diff
@@ -1551,60 +1705,59 @@ int xdp_multiprog__attach(struct xdp_multiprog *mp,
 		xdp_flags |= XDP_FLAGS_DRV_MODE;
 		break;
 	case XDP_MODE_HW:
-		xdp_flags |= XDP_FLAGS_HW_MODE;
+		return -EOPNOTSUPP;
 		break;
 	case XDP_MODE_UNSPEC:
 		break;
 	}
 
-	if (!force)
-		xdp_flags |= XDP_FLAGS_UPDATE_IF_NOEXIST;
-
-	err = bpf_set_link_xdp_fd(ifindex, prog_fd, xdp_flags);
-	if (err == -EEXIST && !(xdp_flags & XDP_FLAGS_UPDATE_IF_NOEXIST)) {
-		/* Program replace didn't work, probably because a program of
-		 * the opposite type is loaded. Let's unload that and try
-		 * loading again.
-		 */
-
-		__u32 old_flags = xdp_flags;
-
-		xdp_flags &= ~XDP_FLAGS_MODES;
-		xdp_flags |= (mode == XDP_MODE_SKB) ? XDP_FLAGS_DRV_MODE :
-			XDP_FLAGS_SKB_MODE;
-		err = bpf_set_link_xdp_fd(ifindex, -1, xdp_flags);
-		if (!err)
-			err = bpf_set_link_xdp_fd(ifindex, prog_fd, old_flags);
-	}
+	err = bpf_set_link_xdp_fd_opts(ifindex, prog_fd, xdp_flags, &opts);
 	if (err < 0) {
 		pr_warn("Error attaching XDP program to ifindex %d: %s\n",
 			ifindex, strerror(-err));
 
 		switch (-err) {
 		case EBUSY:
 		case EEXIST:
-			pr_warn("XDP already loaded on device;"
-				" use --force to replace\n");
+			pr_debug("XDP already loaded on device\n");
 			break;
 		case EOPNOTSUPP:
-			pr_warn("Native XDP not supported;"
-				" try using --skb-mode\n");
+			pr_debug("Native XDP not supported; try using SKB mode\n");
 			break;
 		default:
 			break;
 		}
 		goto err;
 	}
 
-	pr_debug("Loaded %zu programs on ifindex '%d'%s\n",
-		 mp->num_links, ifindex,
-		 mode == XDP_MODE_SKB ? " in skb mode" : "");
+	if (mp)
+		pr_debug("Loaded %zu programs on ifindex '%d'%s\n",
+			 mp->num_links, ifindex,
+			 mode == XDP_MODE_SKB ? " in skb mode" : "");
+	else
+		pr_debug("Detached multiprog on ifindex '%d'%s\n",
+			 ifindex, mode == XDP_MODE_SKB ? " in skb mode" : "");
 
 	return 0;
 err:
 	return err;
 }
 
+int xdp_multiprog__detach(struct xdp_multiprog *mp, int ifindex)
+{
+	int err;
+
+	if (!mp || !mp->is_loaded)
+		return -EINVAL;
+
```

Comment:
Yeah, good question. I think maybe we should just remove the mode parameter from the detach functions? Since the API always asks to detach a specific program, should it matter which mode it is attached in?